src/sync/app.cpp

```diff
@@ -81,11 +95,404 @@ static Optional<AppError> check_for_errors(const Response& response) {
+
+        if (auto error = check_for_errors(response)) {
+            return completion_block({}, error);
+        }
+                    
+        nlohmann::json json;
+        try {
+            json = nlohmann::json::parse(response.body);
+        } catch (const std::exception& e) {
+            return completion_block({}, AppError(make_error_code(JSONErrorCode::malformed_json), e.what()));
+        }
+        
+        try {
+            auto user_api_key = App::UserAPIKey {
+                    ObjectId(value_from_json<std::string>(json, "_id").c_str()),
+                    get_optional<std::string>(json, "key"),
+                    value_from_json<std::string>(json, "name"),
+                    value_from_json<bool>(json, "disabled")
+                };
+            return completion_block(user_api_key, {});
+        } catch (const std::exception& e) {
+            return completion_block({}, AppError(make_error_code(JSONErrorCode::malformed_json), e.what()));
+        }
+    };
+    
+    if (!SyncManager::shared().get_current_user()) {
+        return completion_block({}, AppError(make_custom_error_code(0), "SyncUser is null"));
+    }
+
+    nlohmann::json body = {
+        { "name", name }
+    };
+    
+    parent->m_config.transport_generator()->send_request_to_server({
+        HttpMethod::post,
+        route,
+        parent->m_request_timeout_ms,
+        get_request_headers(true),
+        body.dump()
+    }, handler);
+}
+
+void App::UserAPIKeyProviderClient::fetch_api_key(const realm::ObjectId& id,
+                                                   std::function<void (Optional<UserAPIKey>, Optional<AppError>)> completion_block)
+{
+    REALM_ASSERT(parent);
+    std::string route = util::format("%1/auth/%2/%3", parent->m_base_route, user_api_key_provider_key, id.to_string());
+    
+    auto handler = [completion_block](const Response& response) {
+
+        if (auto error = check_for_errors(response)) {
+            return completion_block({}, error);
+        }
+                    
+        nlohmann::json json;
+        try {
+            json = nlohmann::json::parse(response.body);
+        } catch (const std::exception& e) {
+            return completion_block({}, AppError(make_error_code(JSONErrorCode::malformed_json), e.what()));
+        }
+        
+        try {
+            auto user_api_key = App::UserAPIKey {
+                    ObjectId(value_from_json<std::string>(json, "_id").c_str()),
+                    get_optional<std::string>(json, "key"),
+                    value_from_json<std::string>(json, "name"),
+                    value_from_json<bool>(json, "disabled")
+                };
+            return completion_block(user_api_key, {});
+        } catch (const std::exception& e) {
+            return completion_block({}, AppError(make_error_code(JSONErrorCode::malformed_json), e.what()));
+        }
+    };
+    
+    if (!SyncManager::shared().get_current_user()) {
```

Comment:
The current user doesn't appear to be used anywhere here. It would be better to move the check to where it is actually used. One solution is store it here with the check and pass it as a parameter to `get_request_headers` (this also solves a race condition between time of check and time of use when you fetch it again in that method).

If it really is a precondition to have a current user before calling this method, consider throwing an exception instead of reporting the error to the response handler as it seems a bit weird that the completion handler would report invalid use of this function.

Reply:
You're right, it's not needed here. I will remove it as the only place where its needed is in the `get_request_headers` method